examples/python/Boxes_QPE_example.py

```diff
@@ -0,0 +1,561 @@
+
+
+# ## The Controlled Unitary Stage
+
+# Suppose that we had the following decomposition for $H$ in terms of Pauli strings $P_j$ and complex coefficents $\alpha_j$.
+# 
+# \begin{equation}
+# H = \sum_j \alpha_j P_j\,, \quad ,\, P_j \in \{I, X, Y, Z\}^{\otimes n}
+# \end{equation}
+# 
+# In such a case we could synthesise $U$ as a sequence of Pauli gadgets using pytket's `PauliExpBox`. See the [user manual](https://cqcl.github.io/pytket/manual/manual_circuit.html#pauli-exponential-boxes-and-phase-polynommials) for more discussion.
+# 
+# As an example lets consider a Hamiltonian for diatomic hydrogen $H_2$.
+
+# In[7]:
+
+
+from pytket import Qubit
+from pytket.pauli import Pauli, QubitPauliString
+from pytket.utils import QubitPauliOperator
+
+q = [Qubit(i) for i in range(4)]
+iiii = QubitPauliString(q, [Pauli.I] * 4 )
+iiiz = QubitPauliString(q, [Pauli.I, Pauli.I, Pauli.I, Pauli.Z] )
+iizi = QubitPauliString(q, [Pauli.I, Pauli.I, Pauli.Z, Pauli.I] )
+izii = QubitPauliString(q, [Pauli.I, Pauli.Z, Pauli.I, Pauli.I] )
+ziii = QubitPauliString(q, [Pauli.Z, Pauli.I, Pauli.I, Pauli.I] )
+zizi = QubitPauliString([q[0], q[2]], [Pauli.Z, Pauli.Z])
+iziz = QubitPauliString([q[0], q[3]], [Pauli.Z, Pauli.Z])
+zzii = QubitPauliString([q[0], q[1]], [Pauli.Z, Pauli.Z])
+izzi = QubitPauliString([q[1], q[2]], [Pauli.Z, Pauli.Z])
+yxxy = QubitPauliString(q, [Pauli.Y, Pauli.X, Pauli.X, Pauli.Y])
+xxyy = QubitPauliString(q, [Pauli.X, Pauli.X, Pauli.Y, Pauli.Y])
+yyxx = QubitPauliString(q, [Pauli.Y, Pauli.Y, Pauli.X, Pauli.X])
+xyyx = QubitPauliString(q, [Pauli.X, Pauli.Y, Pauli.Y, Pauli.X])
+ziiz = QubitPauliString(q, [Pauli.Z, Pauli.I, Pauli.I, Pauli.Z])
+iizz = QubitPauliString(q, [Pauli.I, Pauli.I, Pauli.Z, Pauli.Z])
+
+
+# In[8]:
+
+
+qpo_dict = {QubitPauliString(): -0.0596205827603462, iiiz: 0.175759429183197, 
+            iizi: -0.236671176780357, izii: 0.175759429183197,
+            iiiz: -0.236671176780357, zizi: 0.122227149362618, iziz: 0.122227149362618,
+            zzii: 0.170015464396032, izzi: 0.167144319253372,
+            yxxy: 0.0449171698907539, xxyy: -0.0449171698907539, 
+            yyxx: -0.0449171698907539, xyyx: 0.0449171698907539,
+            ziiz: 0.167144319253372, iizz: 0.175703383319070}
+
+
+# In[9]:
+
+
+qpo_h2 = QubitPauliOperator(qpo_dict)
+
+
+# In[10]:
+
+
+from pytket.utils import gen_term_sequence_circuit
+from pytket.passes import DecomposeBoxes
+
+h2_circ = gen_term_sequence_circuit(qpo_h2, Circuit(4))
+h2_circ.name = "U"
+
+render_circuit_jupyter(h2_circ)
+
+
+# We can create a controlled unitary $U$ with a `QControlBox` with $n$ controls
+
+# In[11]:
+
+
+from pytket.circuit import QControlBox
+
+u_box = CircBox(h2_circ)
+
+
+# In[12]:
+
+
+controlled_u = QControlBox(u_box, n=1)
+
+
+# In[13]:
+
+
+test_circ = Circuit(5).add_gate(controlled_u, [0, 1, 2, 3, 4])
+render_circuit_jupyter(test_circ)
+
+
+# ## Putting it all together
+
+# In[14]:
+
+
+def build_phase_est_circuit(n_measurement_qubits: int,
```

Comment:
I'd rewrite as this:
```def build_phase_est_circuit(n_measurement_qubits: int,
                            state_prep_circuit: Circuit,
                            unitary_circuit: Circuit) -> Circuit:
    qpe_circ: Circuit = Circuit()
    measurement_register = qpe_circ.add_q_register("m", n_measurement_qubits)
    state_prep_register = qpe_circ.add_q_register("p", state_prep_circuit.n_qubits)
    
    qpe_circ.add_circuit(state_prep_circuit, list(state_prep_register))
    
    unitary_circuit.name = "U"
    controlled_u = QControlBox(CircBox(unitary_circuit), 1)
    
    for m_qubit in measurement_register:
        qpe_circ.H(m_qubit)
        
    for m_qubit in range(n_measurement_qubits):
        control_qubits: List[Qubit]  = [measurement_register[n_measurement_qubits - m_qubit - 1]] + list(state_prep_qubits)  
        for _ in range(2**m_qubit):
                qpe_circ.add_qcontrolbox(
                    controlled_u, control_qubits
                )
    qpe_circ.add_circbox(CircBox(build_qft_circuit(n_measurement_qubits)).dagger, list(measurement_register))
    qpe_circ.measure_register(measurement_register, "c")
    
    return qpe_circ
```

But also reformat with black, double check it works and add comments. We can use slightly more condensed code here I think

Reply:
Thanks, I've made some of the suggested changes but not all. Let me know if its acceptable.